### ISSUE TITLE
SSL configuration on Dockerfile deployment

### ISSUE THREAD

**Issue by fdoxyz** [dokku/dokku#2113]
**Description of problem:**
I have a `blog` deployed to Dokku v0.5.5 with the default subdomain `blog.site.com` but with an extra url link to the global domain `site.com`. The deployment according to [the Dockerfile docs](http://dokku.viewdocs.io/dokku/deployment/dockerfiles/) when no custom PORT is exposed then 5000 is used and both 80 & 443 are forwarded there. 

Created a certificate using the plugin `dokku certs:generate`. Regular HTTP traffic is handled correctly, but when trying to access using HTTPS I'm getting a 

```
This site can’t provide a secure connection
site.com sent an invalid response.
ERR_SSL_PROTOCOL_ERROR
```

Output of the following commands

- `uname -a`:

`Linux visualcosita.xyz 3.13.0-57-generic #95-Ubuntu SMP Fri Jun 19 09:28:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux`

- `free -m`:
```
                    total       used       free     shared    buffers     cached
Mem:           490        386        104          0         21        159
-/+ buffers/cache:        205        284
Swap:         1023          0       1023
```
- `docker version`:

```
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64
```
- `docker -D info`:

```
Containers: 3
 Running: 2
 Paused: 0
 Stopped: 1
Images: 75
Server Version: 1.11.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 79
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-57-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 490 MiB
Name: visualcosita.xyz
ID: 6B22:ON6Y:TXHF:36EI:CVXE:MDIP:UEPK:QGTZ:XVRY:ONKU:6MOM:YTYA
Docker Root Dir: /var/lib/docker
Debug mode (client): true
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

- `sigil -v`:

`0.4.0`

- `docker run -ti gliderlabs/herokuish:latest herokuish version`:

```
herokuish: 0.3.11
buildpacks:
  heroku-buildpack-clojure   v75
  heroku-buildpack-erlang    fa17af9
  heroku-buildpack-go        v31
  heroku-buildpack-gradle    v17
  heroku-buildpack-grails    v20
  heroku-buildpack-java      v44
  heroku-buildpack-multi     v1.0.0
  heroku-buildpack-nodejs    v89
  heroku-buildpack-php       v100
  heroku-buildpack-play      v26
  heroku-buildpack-python    v80
  heroku-buildpack-ruby      v146
  heroku-buildpack-scala     v67
  buildpack-nginx            v5
```
- `dokku version`:

```
0.5.5
```
- `dokku plugin`:

```
plugn: dev
  00_dokku-standard    0.5.5 enabled    dokku core standard plugin
  20_events            0.5.5 enabled    dokku core events logging plugin
  apps                 0.5.5 enabled    dokku core apps plugin
  build-env            0.5.5 enabled    dokku core build-env plugin
  certs                0.5.5 enabled    dokku core certificate management plugin
  checks               0.5.5 enabled    dokku core checks plugin
  common               0.5.5 enabled    dokku core common plugin
  config               0.5.5 enabled    dokku core config plugin
  docker-options       0.5.5 enabled    dokku core docker-options plugin
  domains              0.5.5 enabled    dokku core domains plugin
  enter                0.5.5 enabled    dokku core enter plugin
  git                  0.5.5 enabled    dokku core git plugin
  logs                 0.5.5 enabled    dokku core logs plugin
  mongo                1.0.0 enabled    dokku mongo service plugin
  named-containers     0.5.5 enabled    dokku core named containers plugin
  nginx-vhosts         0.5.5 enabled    dokku core nginx-vhosts plugin
  plugin               0.5.5 enabled    dokku core plugin plugin
  proxy                0.5.5 enabled    dokku core proxy plugin
  ps                   0.5.5 enabled    dokku core ps plugin
  shell                0.5.5 enabled    dokku core shell plugin
  storage              0.5.5 enabled    dokku core storage plugin
  tags                 0.5.5 enabled    dokku core tags plugin
  tar                  0.5.5 enabled    dokku core tar plugin
```

- `cat /home/dokku/<app>/nginx.conf` (if applicable):

```
server {
  listen      [::]:80;
  listen      80;
  server_name blog.visualcosita.xyz visualcosita.xyz;
  access_log  /var/log/nginx/blog-access.log;
  error_log   /var/log/nginx/blog-error.log;

  location    / {

    gzip on;
    gzip_min_length  1100;
    gzip_buffers  4 32k;
    gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
    gzip_vary on;
    gzip_comp_level  6;

    proxy_pass  http://blog-80;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header Host $http_host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Port $server_port;
    proxy_set_header X-Request-Start $msec;
  }
  include /home/dokku/blog/nginx.conf.d/*.conf;
}

upstream blog-80 {
  server 172.17.0.4:80;
}


server {
  listen      [::]:443;
  listen      443;
  server_name blog.visualcosita.xyz visualcosita.xyz;
  access_log  /var/log/nginx/blog-access.log;
  error_log   /var/log/nginx/blog-error.log;

  location    / {

    gzip on;
    gzip_min_length  1100;
    gzip_buffers  4 32k;
    gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
    gzip_vary on;
    gzip_comp_level  6;

    proxy_pass  http://blog-443;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header Host $http_host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Port $server_port;
    proxy_set_header X-Request-Start $msec;
  }
  include /home/dokku/blog/nginx.conf.d/*.conf;
}

upstream blog-443 {
  server 172.17.0.4:443;
}
```

Environment details (AWS, VirtualBox, physical, etc.):

Running on a DigitalOcean droplet

How was dokku installed?:

Installed old 3.x/4.x version using the `bootstrap.sh` but updated using `apt-get` to v0.5.5

Steps to Reproduce:
1. Deploy app
2. Create a link for the app to the global url
3. Create self-signed certificates `dokku certs:generate blog`
4. Access the site via HTTP
5. Access the site via HTTPS

Actual Results:
HTTP requests work and HTTPS request don't

Expected Results:
Both should work, with a self-signed certificate anyways

Additional info:
The repo of the project is here to clone and reproduce https://github.com/fdoxyz/visualcosita

**Comment by fdoxyz** [comment]
Trying to hack a workaround for this issue, by handling both port 80 & 443 traffic in the same port, I've come to notice that the `:443` forwarding/upstream in `nginx.conf` was only created when I tried to use the certs plugin. The only forwarding in the conf is of port 80.

I did a clean install (including droplet to fresh ubuntu 14.04) of Dokku and deployed with the Dockerfile, this should forward ports 80 & 443 to default 5000 (when not specified in the Dockerfile) but it doesn't. I'm not well versed in nginx conf templating but maybe this is part of the bug. This is the generated `nginx.conf` without any cert plugin tampering:

```
server {
  listen      [::]:80;
  listen      80;
  server_name blog.visualcosita.xyz visualcosita.xyz;
  access_log  /var/log/nginx/blog-access.log;
  error_log   /var/log/nginx/blog-error.log;


  location    / {

    gzip on;
    gzip_min_length  1100;
    gzip_buffers  4 32k;
    gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
    gzip_vary on;
    gzip_comp_level  6;

    proxy_pass  http://blog;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header Host $http_host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Port $server_port;
    proxy_set_header X-Request-Start $msec;
  }
  include /home/dokku/blog/nginx.conf.d/*.conf;


}


upstream blog {
  server 172.17.0.2:5000;
}
```

**Comment by michaelshobbs** [comment]
By default when there is an SSL cert present, dokku will generate an nginx config that forwards port 80 traffic to SSL on 443 and then upstream to the docker container. It seems to work using the repo you included and a self signed cert. Have a look at what I attempted below and let me know what is isn't working for your env.

create app and cert
```
$ ssh dokku.me apps:create test
Creating test... done
Connection to dokku.me closed.
$ ssh dokku.me certs:generate test
Generating RSA private key, 2048 bit long modulus
.......................................................+++
........................................+++
e is 65537 (0x10001)
writing RSA key
You are about to be asked to enter information that will be incorporated
into your certificate request.
...snip...
-----> Installing certificate and key...
-----> The following is a certificate signing request that can be used
-----> to generate an 'officially' signed SSL certificate for test at
-----> by a CA of your choosing.
-----BEGIN CERTIFICATE REQUEST-----
...snip...
-----END CERTIFICATE REQUEST-----
Connection to dokku.me closed.
```

deploy app
```
$ git push dokku@dokku.me:test master
Pseudo-terminal will not be allocated because stdin is not a terminal.
Counting objects: 388, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (204/204), done.
Writing objects: 100% (388/388), 987.96 KiB | 0 bytes/s, done.
Total 388 (delta 132), reused 388 (delta 132)
-----> Cleaning up...
-----> Building test from dockerfile...
-----> Setting config vars
       DOKKU_DOCKERFILE_CMD: CMD ["node", "server.js"]
remote: Sending build context to Docker daemon 1.022 MB
Step 1 : FROM mhart/alpine-node:4
...snip
Removing intermediate container 125081e00bb2
Successfully built 0b47e0b9fe59
-----> Releasing test (dokku/test:latest)...
-----> Deploying test (dokku/test:latest)...
-----> Attempting to run scripts.dokku.predeploy from app.json (if defined)
-----> No Procfile found in app image
-----> DOKKU_SCALE file not found in app image. Generating one based on Procfile...
-----> New DOKKU_SCALE file generated
=====> web=1
-----> Running pre-flight checks
       For more efficient zero downtime deployments, create a file CHECKS.
       See http://dokku.viewdocs.io/dokku/checks-examples.md for examples
       CHECKS file not found in container: Running simple container check...
-----> Waiting for 10 seconds ...
-----> Default container check successful!
-----> Running post-deploy
-----> Attempting to run scripts.dokku.postdeploy from app.json (if defined)
=====> renaming container (ac7afe3ff390) admiring_curie to test.web.1
-----> Creating new /home/dokku/test/VHOST...
-----> Setting config vars
       DOKKU_NGINX_PORT: 80
-----> Setting config vars
       DOKKU_NGINX_SSL_PORT: 443
-----> Configuring test.dokku.me...(using built-in template)
-----> Creating https nginx.conf
-----> Running nginx-pre-reload
       Reloading nginx
-----> Setting config vars
       DOKKU_APP_RESTORE: 1
=====> Application deployed:
       https://test.dokku.me

To dokku@dokku.me:test
 * [new branch]      master -> master
```
add alternate domain
```
$ ssh dokku.me domains:add test dokku.me
-----> Added dokku.me to test
-----> Configuring dokku.me...(using built-in template)
-----> Configuring test.dokku.me...(using built-in template)
-----> Creating https nginx.conf
-----> Running nginx-pre-reload
       Reloading nginx
Connection to dokku.me closed.
```

/home/dokku/$APP/nginx.conf
```
$ cat /home/dokku/test/nginx.conf

server {
  listen      [::]:80;
  listen      80;
  server_name test.dokku.me dokku.me;
  access_log  /var/log/nginx/test-access.log;
  error_log   /var/log/nginx/test-error.log;


  return 301 https://$host:443$request_uri;

}

server {
  listen      [::]:443 ssl spdy;
  listen      443 ssl spdy;
  server_name test.dokku.me;
  server_name test.dokku.me dokku.me;
  access_log  /var/log/nginx/test-access.log;
  error_log   /var/log/nginx/test-error.log;

  ssl_certificate     /home/dokku/test/tls/server.crt;
  ssl_certificate_key /home/dokku/test/tls/server.key;

  keepalive_timeout   70;
  add_header          Alternate-Protocol  443:npn-spdy/2;
  location    / {

    gzip on;
    gzip_min_length  1100;
    gzip_buffers  4 32k;
    gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
    gzip_vary on;
    gzip_comp_level  6;

    proxy_pass  http://test;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header Host $http_host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Port $server_port;
    proxy_set_header X-Request-Start $msec;
  }
  include /home/dokku/test/nginx.conf.d/*.conf;

}


upstream test {
  server 172.17.0.6:5000;
}
```
nginx answers on port 80 and redirects to 443
```
$ curl -skLi http://test.dokku.me
HTTP/1.1 301 Moved Permanently
Server: nginx
Date: Sun, 17 Apr 2016 20:57:43 GMT
Content-Type: text/html
Content-Length: 178
Connection: keep-alive
Location: https://test.dokku.me:443/

HTTP/1.1 200 OK
Server: nginx
Date: Sun, 17 Apr 2016 20:57:43 GMT
Content-Type: text/html
Content-Length: 6586
Connection: keep-alive
Vary: Accept-Encoding
cache-control: max-age=10800
Etag: "48-6586-1460926404000"
Last-Modified: Sun, 17 Apr 2016 20:53:24 GMT
Alternate-Protocol: 443:npn-spdy/2
```

nginx answers on port 443 as well
```
$ curl -skLi https://test.dokku.me
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 17 Apr 2016 20:57:35 GMT
Content-Type: text/html
Content-Length: 6586
Connection: keep-alive
Vary: Accept-Encoding
cache-control: max-age=10800
Etag: "48-6586-1460926404000"
Last-Modified: Sun, 17 Apr 2016 20:53:24 GMT
Alternate-Protocol: 443:npn-spdy/2
```

alternate domains
```
$ curl -skLi http://dokku.me
HTTP/1.1 301 Moved Permanently
Server: nginx
Date: Sun, 17 Apr 2016 21:04:17 GMT
Content-Type: text/html
Content-Length: 178
Connection: keep-alive
Location: https://dokku.me:443/

HTTP/1.1 200 OK
Server: nginx
Date: Sun, 17 Apr 2016 21:04:17 GMT
Content-Type: text/html
Content-Length: 6586
Connection: keep-alive
Vary: Accept-Encoding
cache-control: max-age=10800
Etag: "48-6586-1460926404000"
Last-Modified: Sun, 17 Apr 2016 20:53:24 GMT
Alternate-Protocol: 443:npn-spdy/2

$ curl -skLi https://dokku.me
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 17 Apr 2016 21:04:22 GMT
Content-Type: text/html
Content-Length: 6586
Connection: keep-alive
Vary: Accept-Encoding
cache-control: max-age=10800
Etag: "48-6586-1460926404000"
Last-Modified: Sun, 17 Apr 2016 20:53:24 GMT
Alternate-Protocol: 443:npn-spdy/2
```

**Comment by fdoxyz** [comment]
Maybe it was because I was using an install coming from 0.3.x upgraded multiple times to 0.4.x and then 0.5.x until the current version. I also removed the certs and regenerated them a couple of times (with warnings every now and then). Either one of the previous scenarios caused my problems or the order of the commands I was using didn't work.

Tried again after destroying the app, following your process and the nginx.conf file was generated correctly on the fresh install.

I thought port 443 was forwarded by default even without setting up certs using the plugin, now I know **it will only be set after setting up the certs with the plugin**.

The only note I can add is that the certs plugin is **not listed** in the `dokku help` command, even though **it is** listed as a core plugin when calling `dokku plugin`, not sure if this is intended but worth noting.

Thanks a lot for your help, amazing work on the project :)
